OpenRA.Game/Actor.cs

```diff
@@ -118,6 +121,16 @@ internal Actor(World world, string name, TypeDictionary initDict)
 			visibilityModifiers = TraitsImplementing<IVisibilityModifier>().ToArray();
 			defaultVisibility = Trait<IDefaultVisibility>();
 			Targetables = TraitsImplementing<ITargetable>().ToArray();
+			targetablePositions = TraitsImplementing<ITargetablePositions>().ToArray();
+			this.World.AddFrameEndTask(w =>
+			{
+				// Caching this in a AddFrameEndTask, because trait construction order might cause problems if done directly at creation time
+				if (!this.Info.HasTraitInfo<IMoveInfo>() && targetablePositions.All(tp => tp.AlwaysEnabled))
+				{
+					onlyStaticTargetablePositions = true;
+					staticTargetablePositions = targetablePositions.SelectMany(tp => tp.TargetablePositions(this)).ToArray();
+				}
+			});
```

Comment:
If we're going to deal in providing actual positions, I wonder if you can streamline the caching to focus on just that

- Define `IEnumerable<WPos> targetablePositions;`
- Set `targetablePositions = TraitsImplementing<ITargetablePositions>().ToArray().Where(Exts.IsTraitEnabled).HandleTheAnyCheckThatOtherwiseReturnsTheCenterSomehow().SelectMany(tp => tp.TargetablePositions(this));`
- In the frame task, if the "won't ever change" condition is met, set `targetablePositions = targetablePositions.ToArray()`;

You could then expose `targetablePositions` as a public property - and you don't need the other fields to track stuff or the method.

Reply:
If I'm drinking the kool-aid and the above suggestion is crazy - could you at least remove `onlyStaticTargetablePositions` and simply do null checks on `staticTargetablePositions` instead?

Reply:
I wouldn't call it crazy, but besides `HandleTheAnyCheckThatOtherwiseReturnsTheCenterSomehow()` there's also the issue that if the "won't ever change" condition is *not* met, we'd need to update `TargetablePositions` when called, I don't see how that's supposed to work if we drop both the boolean and the method.